index.html

```diff
@@ -2435,37 +2435,30 @@ <h2>Cryptographic Suites</h2>
 verification of the data.
         </li>
         <li>
-          The specification MUST define a <dfn>suite selection algorithm</dfn> that accepts a [=JSON
-          object=] |proof| and uses the properties described in [[[#proofs]]] to return either
-          failure or a [=VC cryptographic suite=] [[Infra]] [=struct=]. This algorithm SHOULD be
-          registered in [[[VC-SPECS]]]. A
-          <dfn class="export">VC cryptographic suite</dfn> [=struct=] has the following
-          [=struct/items=].
-          <dl data-dfn-for="cryptographic suite">
-            <dt><dfn>transform</dfn></dt>
-            <dd>A [=transformation algorithm=] that takes a [=JSON object=] as input and returns a
-            [=JSON object=].</dd>
-            <dt><dfn>hash</dfn></dt>
-            <dd>A [=hashing algorithm=] that takes a [=JSON object=] as input and returns a [=byte
-            sequence=].</dd>
-            <dt><dfn>proof creation options</dfn></dt>
+The specification MUST define a <dfn class="export"
+data-lt="cryptosuite instantiation algorithm">data integrity cryptographic
+suite instantiation algorithm</dfn> that accepts a set of options
+([=map=] |options|) and returns a [=cryptosuite instance=]
+([=struct=] |cryptosuite|). This algorithm SHOULD be listed in [[[VC-SPECS]]].
+A <dfn class="export" data-lt="cryptosuite instance">data integrity
+cryptographic suite instance</dfn> [=struct=] has the following
+[=struct/items=].
+          <dl data-dfn-for="cryptosuite instance">
+            <dt><dfn>createProof</dfn></dt>
             <dd>
-              <p>A [=struct=] type that needs to be supplied to the [=cryptographic suite/serialize
-              proof=] operation.</p>
-              <p class="note">To sign data, cryptographic suites need to be able to use a private
-              key. This private key will often be implied by a field in the [=proof creation
-              options=], but it is often a better design to avoid passing the key directly. Passing
-              a key identifier or an asynchronous signing algorithm allows implementations to take
-              advantage of hardware security modules (HSMs) to get better security.</p>
+An algorithm that takes an [=unsecured data document=] ([=map=]
+|unsecuredDocument|) and proof options ([=map=] |options|) as input, and
+produces a [=data integrity proof=] ([=map=]) or an error.
+            </dd>
+            <dt><dfn>verifyProof</dfn></dt>
+            <dd>
+An algorithm that takes a [=secured data document=] ([=map=]
+|securedDocument|) as input, and produces a verification result
```

Comment:
I'd usually use "returns" for an algorithm instead of "produces" or "outputs", but I also don't insist on it.

```suggestion
|securedDocument|) as input, and returns a verification result
```

Reply:
I was trying to avoid "returns" as we haven't defined an error structure to return; we depend on functions throwing when there are errors/exceptions. It's debatable whether or not that is a "return" value or an exception value since programming environments typically have different syntax for handling each path.

I admit to having struggled with this particular part of documenting lower-level functions... I don't want to insist that developers use return structures for errors (because languages that prefer throwing exceptions exist). "Produces" feels like it gives the implementer more latitude to make the choice, where "returns" feels like we're taking exception handling off of the table.

I know that's a bit esoteric, but that's the current approach that many of the algorithms have taken. Changing it to "return" would require a rewrite to most/all of the algorithms. Do we have a best practices wrt. this particular concern at W3C?

Reply:
I think the Infra spec strives to document whatever best practices we have, but it's fine for WGs to override that in particular cases. Note that whatever you write, you're not insisting that developers use any particular structures for errors, due to https://infra.spec.whatwg.org/#algorithm-conformance:~:text=Conformance%20requirements%20phrased%20as%20algorithms%20or%20specific%20steps%20may%20be%20implemented%20in%20any%20manner%2C%20so%20long%20as%20the%20end%20result%20is%20equivalent. The  most egregious example I know of is https://html.spec.whatwg.org/multipage/infrastructure.html#starting-a-new-parallel-queue, which is specified as a busy wait. But again, I'm not insisting on changing this.